sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/circuitBreaker/GlobalPartitionEndpointManagerForCircuitBreaker.java

```diff
@@ -0,0 +1,509 @@
+        checkNotNull(collectionResourceId, "Argument 'collectionResourceId' cannot be null!");
+
+        PartitionKeyRangeWrapper partitionKeyRangeWrapper = new PartitionKeyRangeWrapper(partitionKeyRange, collectionResourceId);
+
+        AtomicBoolean isFailoverPossible = new AtomicBoolean(true);
+        AtomicBoolean isFailureThresholdBreached = new AtomicBoolean(false);
+
+        String collectionLink = getCollectionLink(request);
+
+        this.partitionKeyRangeToLocationSpecificUnavailabilityInfo.compute(partitionKeyRangeWrapper, (partitionKeyRangeWrapperAsKey, partitionLevelLocationUnavailabilityInfoAsVal) -> {
+
+            if (partitionLevelLocationUnavailabilityInfoAsVal == null) {
+                partitionLevelLocationUnavailabilityInfoAsVal = new PartitionLevelLocationUnavailabilityInfo();
+            }
+
+            isFailureThresholdBreached.set(partitionLevelLocationUnavailabilityInfoAsVal.handleException(
+                partitionKeyRangeWrapperAsKey,
+                collectionLink,
+                failedLocation,
+                request.isReadOnlyRequest()));
+
+            if (isFailureThresholdBreached.get()) {
+
+                UnmodifiableList<URI> applicableEndpoints = request.isReadOnlyRequest() ?
+                    this.globalEndpointManager.getApplicableReadEndpoints(request.requestContext.getExcludeRegions()) :
+                    this.globalEndpointManager.getApplicableWriteEndpoints(request.requestContext.getExcludeRegions());
+
+                isFailoverPossible.set(
+                    partitionLevelLocationUnavailabilityInfoAsVal.areLocationsAvailableForPartitionKeyRange(applicableEndpoints));
+            }
+
+            request.requestContext.setLocationToLocationSpecificHealthContext(partitionLevelLocationUnavailabilityInfoAsVal.regionToLocationSpecificHealthContext);
+            return partitionLevelLocationUnavailabilityInfoAsVal;
+        });
+
+        // set to true if and only if failure threshold exceeded for the region
+        // and if failover is possible
+        // a failover is only possible when there are available regions left to fail over to
+        if (isFailoverPossible.get()) {
+            return;
+        }
+
+        if (logger.isWarnEnabled()) {
+            logger.warn("It is not possible to mark region {} as Unavailable for partition key range {}-{} and collection rid {} " +
+                    "as all regions will be Unavailable in that case, will remove health status tracking for this partition!",
+                this.globalEndpointManager.getRegionName(
+                    failedLocation, request.isReadOnlyRequest() ? OperationType.Read : OperationType.Create),
+                partitionKeyRange.getMinInclusive(),
+                partitionKeyRange.getMaxExclusive(),
+                collectionResourceId);
+        }
+
+        // no regions to fail over to
+        this.partitionKeyRangeToLocationSpecificUnavailabilityInfo.remove(partitionKeyRangeWrapper);
```

Comment:
question - for single master account, does this means the partition circular breaking is disabled? (since write operation has no failover region?)

Reply:
Correct.